### DISCUSSION TITLE
Return tuple of values from C or C++ to Python

### DISCUSSION THREAD

**Discussion by AutoPilotjn** [micropython/micropython#10529] [General]
Hi,
Is there a way to return a tuple (of doubles) from C or C++ to within a Python script ?
I read the MicroPython external C modules/Basic example (cmodule). I also read source codes such as 'upython-1.19.1/py/runtime.c'. However I could not find out a way to return tuple of doubles.
it looks like:
`mp_obj_tuple_t *myTuple = mp_obj_new_tuple(x, NULL);
`
allocates x integers, not doubles.
Is the structure myTuple allocated automatically by the 'mp_obj_new_tuple()' function ? Or shall it be allocated manually in the source code ? In this later case, how ?
More generally, how to return tuple of variables of different types ?
Thanks for  help

**Comment by dlech** [comment]
you need to create a C array of `mp_obj_t`, create float objects and store them in the array and then pass that to `mp_obj_new_tuple()`.

**Comment by AutoPilotjn** [comment]
Yes !
As simple as:
```C++
    mp_obj_t get_quaternions() {
        mp_obj_t tupleQ[4];
        tupleQ[0] = mp_obj_new_float(myFilter.getQuaternionQ0());
        tupleQ[1] = mp_obj_new_float(myFilter.getQuaternionQ1());
        tupleQ[2] = mp_obj_new_float(myFilter.getQuaternionQ2());
        tupleQ[3] = mp_obj_new_float(myFilter.getQuaternionQ3());
        return mp_obj_new_tuple(4, tupleQ);
}
```
and within the python script:
```python
    import MadgwickAHRS
    t_tuple_quaternion_float: TypeAlias = tuple[float, float, float, float]
    def read_quaternions(self) -> t_tuple_quaternion_float:
        """Get quaternion values: (q0, q1, q2 & q3)"""
        return (
            MadgwickAHRS.get_quaternions()
        )
```
Thanks

**Reply by jgpeiro** [reply]
Create micropython wrappers is always a difficult task. There are few [tools](https://forum.micropython.org/viewtopic.php?t=7427&p=42351) and [documents ](https://micropython-usermod.readthedocs.io/en/latest/) that explains how to do it...
But with Copilot makes it much easier:
 
![image](https://user-images.githubusercontent.com/6277497/213884336-0affca88-0af6-4c18-87bd-43a22cbd33f2.png)


Question:
Is "t_tuple_quaternion_float: TypeAlias = tuple[float, float, float, float]" supported in micropython?

**Reply by jimmo** [reply]
@jgpeiro 

> Is "t_tuple_quaternion_float: TypeAlias = tuple[float, float, float, float]" supported in micropython?

This is PEP613, added in CPython 3.10.

Not supported in MicroPython, see https://docs.micropython.org/en/latest/differences/python_310.html

**Reply by AutoPilotjn** [reply]
Correct, TypeAlias is not implemented in MicroPython. However there are two possibilities:
 - add a typing module as described here:
[https://github.com/micropython/micropython-lib/pull/584]
 - or remove the typing hints in the source code through AST analysis as described here:
[https://github.com/orgs/micropython/discussions/9327#discussioncomment-4121023](url) 
These two options bring all innovative features of the Python ecosystem on the IDE before running on the target. The latest option has no limitation in removing what might look "esoteric type hints".

**Comment by AutoPilotjn** [comment]
It is possible to return a tuple of various types from C or C++ to within a MicroPython script.
C or C++ (filterAHRS module):
```C
mp_obj_t get_tuple() {
	mp_obj_t tupleQ[3];
	char testChar[5] = "Test";
	tupleQ[0] = mp_obj_new_int(1234);
	tupleQ[1] = mp_obj_new_float(9876.1234);
	tupleQ[2] = mp_obj_new_str(testChar, 5);
	return mp_obj_new_tuple(3, tupleQ);
}
```
Python script:
```python
import filterAHRS
def read_tuple(self) -> tuple[int, float, str]:
        """Get various types from typle"""
        int_number, float_number, string = filterAHRS.get_tuple()
        print(f"integer: {int_number}, float: {float_number}, string: {string}")
```
Works like a charm !
